harvester_e2e_tests/integration/test_backup_restore.py

```diff
@@ -0,0 +1,803 @@
+        ip = _wait_for_vm_ip(endpoint, api_client, vm_name, wait_timeout)
+        with host_shell.login(ipaddr=endpoint_ip, jumphost=True) as host, \
+             vm_shell.login(ipaddr=ip,
+                            username=USERNAME,
+                            password=PASSWORD,
+                            pkey=private_key,
+                            jumphost=host.client) as vm:
+            intact = (_check_file_integrity(vm, 'test-0') and
+                      _check_file_integrity(vm, 'test-1'))
+            assert intact, (
+                'Restore failed, "test-0" file not found or md5 mismatch.')
+
+        if need_cleanup:
+            api_client.vms.delete(vm_name)
+
+    def test_restore_replace_current_vm(self, endpoint, api_client,
+                                        vm_backups, wait_timeout, host_shell,
+                                        vm_shell, need_cleanup, ssh_keypair,
+                                        endpoint_ip, vm_name):
+        backup_name = vm_backups[1]['metadata']['name']
+        spec = api_client.backups.RestoreSpec.for_existing(delete_volumes=True)
+
+        _wait_for_vm_stop(api_client, wait_timeout, vm_name)
+        code, data = api_client.backups.restore(backup_name,
+                                                spec)
+        assert 201 == code, (
+            f'Failed to restore backup with current VM replaced, {data}')
+
+        _, private_key = ssh_keypair
+        _wait_for_vm_ready(api_client, wait_timeout, vm_name)
+        ip = _wait_for_vm_ip(endpoint, api_client, vm_name, wait_timeout)
+        with host_shell.login(ipaddr=endpoint_ip, jumphost=True) as host, \
+             vm_shell.login(ipaddr=ip,
+                            username=USERNAME,
+                            password=PASSWORD,
+                            pkey=private_key,
+                            jumphost=host.client) as vm:
+            intact = (_check_file_integrity(vm, 'test-0') and
+                      _check_file_integrity(vm, 'test-1'))
+            assert intact, (
+                'Restore failed, "test-0" file not found or md5 mismatch.')
+
+        if need_cleanup:
+            api_client.vms.delete(vm_name)
+
+
+@pytest.mark.p0
+@pytest.mark.backupnfs
+class TestDeleteBackupNFS:
+    def test_delete_first_backup(self, endpoint, api_client,
+                                 vm_name, restored_vm_name, wait_timeout,
+                                 vm_backups, ssh_keypair, need_cleanup,
+                                 host_shell, vm_shell, endpoint_ip):
+        backup_to_delete = vm_backups[0]['metadata']['name']
+        code, data = api_client.backups.delete(backup_to_delete)
+        assert 200 == code, (
+            f'Failed to delete backup: {data}')
+
+        _, private_key = ssh_keypair
+        ip = _wait_for_vm_ip(endpoint, api_client, vm_name, wait_timeout)
+        _wait_for_vm_ready(api_client, wait_timeout, vm_name)

```

Comment:
When a VM got IP which means it should be ready.
we can check `AgentConnected` in VMI's last conditions and `interfaces` in VMI's status.

Ref:
https://github.com/harvester/tests/blob/13972cd5446a0e7e18c0d82d8ba15e664c61da68/harvester_e2e_tests/integration/test_vm_functions.py#L536-L560